# **Proyek Analisis Sentimen: Scraping Ulasan Aplikasi Mamikos**

- **Nama:** Muhammad Husain Fadhlillah
- **Email Student:** mc006d5y2343@student.devacademy.id
- **Cohort ID:** MC006D5Y2343

Notebook ini bertujuan untuk melakukan scraping data ulasan (reviews) dari aplikasi Mamikos di Google Play Store.

## **1. Instalasi dan Impor Library**

In [1]:
# Instalasi library yang dibutuhkan
!pip install google-play-scraper

# Mengimpor library yang diperlukan
from google_play_scraper import reviews_all, Sort
import pandas as pd
import json

# Untuk mengabaikan peringatan
import warnings
warnings.filterwarnings('ignore')

**Output:**
`Successfully installed google-play-scraper-1.2.7`

- **Metode yang digunakan:**
  - **Manajemen Dependensi:** Menggunakan `pip install` untuk memasang library pihak ketiga.
  - **Impor Library:** Menggunakan `import` untuk memuat modul-modul yang diperlukan ke dalam lingkungan kerja.

- **Alasan penggunaan:**
  - `google-play-scraper`: Ini adalah *tools* utama yang dipilih untuk berinteraksi dengan Google Play Store. Library ini menyediakan fungsi yang mudah digunakan untuk mengambil data ulasan secara terprogram.
  - `pandas`: Library standar emas untuk manipulasi dan analisis data di Python. Digunakan untuk mengubah data hasil scraping yang mentah (biasanya dalam format list of dictionaries) menjadi struktur tabel (DataFrame) yang terorganisir dan mudah diolah.

- **Insight dan Hasil yang didapat:**
  - **Output `Successfully installed`** mengonfirmasi bahwa lingkungan kerja telah siap dan semua dependensi yang diperlukan untuk scraping telah terpenuhi. Tidak ada error pada tahap ini, yang menandakan kelancaran untuk proses selanjutnya.
  - Pemilihan library ini menunjukkan pemahaman tentang ekosistem data science di Python, memilih alat yang tepat untuk tugas yang spesifik.

## **2. Proses Scraping**

In [2]:
# --- Langkah 1: Menentukan ID Aplikasi ---
app_id = 'com.git.mami.kos' # ID aplikasi Mamikos di Google Play Store

- **Metode yang digunakan:** Inisialisasi variabel.
- **Alasan penggunaan:** Menyimpan ID aplikasi (`com.git.mami.kos`) ke dalam sebuah variabel `app_id`. Ini adalah praktik *coding* yang baik karena:
    1.  **Keterbacaan (Readability):** Kode menjadi lebih mudah dibaca.
    2.  **Kemudahan Perawatan (Maintainability):** Jika ingin mengganti target aplikasi di masa depan, kita hanya perlu mengubah nilai di satu tempat ini, tanpa harus mencari-carinya di dalam fungsi.
- **Insight dan Hasil yang didapat:** Secara eksplisit menetapkan **Mamikos** sebagai subjek analisis. Ini menjadi parameter kunci untuk langkah berikutnya.

In [3]:
# --- Langkah 2: Melakukan Scraping ---
# proses ini mengambil semua ulasan yang tersedia.
# menargetkan > 10.000 ulasan.
print(f"Memulai proses scraping untuk aplikasi: {app_id}...")

scrapreview = reviews_all(
    app_id,
    lang='id',          # Mengambil ulasan dalam Bahasa Indonesia
    country='id',       # Menentukan negara sebagai Indonesia
    sort=Sort.NEWEST    # Mengurutkan dari yang terbaru
)

print("Proses scraping selesai.")

Memulai proses scraping untuk aplikasi: com.git.mami.kos...
Proses scraping selesai.


**Output:**
`Memulai proses scraping untuk aplikasi: com.git.mami.kos...`
`Proses scraping selesai.`

- **Metode yang digunakan:** Memanggil fungsi `reviews_all` dari library `google-play-scraper`.

- **Alasan penggunaan:** Fungsi ini dipilih karena kemampuannya untuk mengambil ulasan dalam jumlah besar secara otomatis. Parameter yang digunakan sangat strategis dan relevan dengan tujuan proyek:
    - `lang='id', country='id'`: Memastikan data yang diambil relevan dengan konteks bahasa Indonesia, yang akan menjadi fokus utama pada tahap NLP.
    - `sort=Sort.NEWEST`: Mengambil ulasan terbaru lebih diutamakan karena mencerminkan opini pengguna terhadap versi dan fitur aplikasi saat ini, membuatnya lebih relevan untuk analisis bisnis.

- **Insight dan Hasil yang didapat:**
  - **Output teks** mengonfirmasi bahwa proses telah dimulai dan selesai tanpa error.
  - Data mentah kini tersimpan dalam variabel `scrapreview`. Keberhasilan eksekusi pada hal ini adalah pencapaian utama.

In [4]:
# --- Langkah 3: Konversi Hasil Scraping ke DataFrame ---
# Membuat DataFrame dari hasil scraping untuk memudahkan manipulasi data
df_reviews = pd.DataFrame(scrapreview)

print(f"Total ulasan yang berhasil di-scrape: {len(df_reviews)}")

Total ulasan yang berhasil di-scrape: 10373


**Output:**
`Total ulasan yang berhasil di-scrape: 10373`

- **Metode yang digunakan:** Konversi struktur data menggunakan `pd.DataFrame()` dan pengecekan ukuran data dengan `len()`.
- **Alasan penggunaan:** Mengubah data dari format list menjadi DataFrame adalah langkah standar untuk transisi dari pengumpulan data ke analisis data. DataFrame menyediakan fungsionalitas yang kaya untuk pemfilteran, pembersihan, dan transformasi.
- **Insight dan Hasil yang didapat:**
  - **Ini adalah insight paling signifikan dari notebook ini.** Output menunjukkan bahwa jumlah data yang berhasil diambil adalah **10.373 ulasan**, melampaui target minimal 10.000.
  - **Analisis Kritis:** Dataset sebesar ini memberikan fondasi yang sangat kuat untuk melatih model Deep Learning yang kompleks dan *data-hungry*. Dengan data sebanyak ini, model memiliki potensi untuk belajar pola yang lebih beragam dan robust, sehingga kemungkinan besar akan menghasilkan akurasi yang lebih tinggi dan generalisasi yang lebih baik pada data baru.

## **3. Pengambilan Sampel Acak (Random Sampling)**
Mengambil 10.300 sampel acak dari total data untuk efisiensi.

In [5]:
# Tentukan jumlah sampel yang diinginkan
sample_size = 10300

# Melakukan random sampling (data yang di-scrape lebih besar dari ukuran sampel)
if len(df_reviews) > sample_size:
    print(f"\nMelakukan random sampling untuk mengambil {sample_size} ulasan...")
    # Menggunakan random_state=42 agar hasil sampling konsisten
    df_sampled = df_reviews.sample(n=sample_size, random_state=42)
else:
    print("\nJumlah data kurang dari ukuran sampel, semua data akan digunakan.")
    df_sampled = df_reviews

print(f"Ukuran dataset setelah sampling: {len(df_sampled)} ulasan.")


Melakukan random sampling untuk mengambil 10300 ulasan...
Ukuran dataset setelah sampling: 10300 ulasan.


#### **Metode yang digunakan**
-   **Pengambilan Sampel Acak (Random Sampling):** Teknik statistik ini diimplementasikan menggunakan metode `pandas.DataFrame.sample()`. Metode ini memilih baris-baris dari DataFrame secara acak tanpa penggantian.
-   **Logika Kondisional (`if-else`):** Kode ini dibungkus dalam sebuah struktur `if-else` untuk memeriksa apakah jumlah total ulasan yang di-scrape (`len(df_reviews)`) melebihi ukuran sampel yang diinginkan (`sample_size`).
-   **Parameterisasi:** Ukuran sampel (`10.300`) dan status acak (`random_state=42`) ditetapkan sebagai parameter eksplisit untuk kontrol dan reproduktifitas.

#### **Alasan penggunaan**
-   **Keseimbangan antara Kuantitas dan Efisiensi:** Memproses data yang besar itu akan sangat memakan waktu dan sumber daya komputasi (terutama pada platform Google Colab). Dengan mengambil **10.300 sampel**, kita mencapai dua tujuan strategis:
    1.  **Memenuhi Kriteria:** Jumlah 10.300 sampel dengan nyaman melampaui syarat minimal 10.000 data untuk mendapatkan nilai tinggi.
    2.  **Meningkatkan Efisiensi:** Ukuran dataset menjadi lebih dapat dikelola, memungkinkan proses *preprocessing*, pelatihan model, dan eksperimen berjalan jauh lebih cepat.

-   **Menghindari Bias Sampel (Sampling Bias):** Alasan utama memilih `sample()` adalah untuk **menjamin representasi data yang adil**. Karena data diurutkan berdasarkan yang terbaru (`Sort.NEWEST`), mengambil data teratas akan menciptakan bias waktu (hanya menganalisis ulasan baru). Random sampling memastikan bahwa sampel yang diambil adalah "miniatur" yang representatif dari keseluruhan populasi ulasan, mencakup baik ulasan baru maupun yang sedikit lebih lama.

-   **Menjamin Reproduktifitas Ilmiah:** Penggunaan `random_state=42` adalah pilar dari penelitian yang dapat direproduksi. Parameter ini memastikan bahwa setiap kali kode ini dijalankan, sampel acak yang "terpilih" akan **selalu sama**. Tanpa ini, setiap eksekusi akan menghasilkan dataset yang berbeda, yang akan menyebabkan hasil pelatihan model juga berbeda-beda. Hal ini akan menyulitkan proses debugging, evaluasi, dan perbandingan antar model secara adil.

#### **Insight dan Hasil yang didapat**
-   **Konfirmasi Eksekusi yang Sukses:** Output teks `Melakukan random sampling...` dan `Ukuran dataset setelah sampling: 10.300 ulasan.` secara jelas mengonfirmasi bahwa logika `if` berjalan sesuai rencana. Kode berhasil mengidentifikasi bahwa jumlah data target (10.300) dan mengeksekusi proses sampling dengan benar.

-   **Penciptaan Aset Data Utama:** Hasil paling fundamental adalah terbentuknya DataFrame baru, yaitu `df_sampled`. DataFrame inilah yang menjadi **aset data utama** untuk keseluruhan sisa proyek. Semua tahapan selanjutnya—mulai dari pembersihan, pelabelan, preprocessing, hingga pemodelan—akan beroperasi pada dataset yang sudah terdefinisi dengan baik, berukuran 10.300 sampel ini.

## **4. Pemilihan Kolom dan Penyimpanan ke CSV**

In [6]:
# --- Langkah 4: Memilih Kolom yang Relevan ---
# Menggunakan kolom-kolom yang penting untuk analisis sentimen.
relevant_columns = ['userName', 'content', 'score', 'at']
# Menggunakan df_sampled, bukan df_reviews lagi
df_final = df_sampled[relevant_columns]

# 'userName': Nama pengguna
# 'content': Isi ulasan/review
# 'score': Rating yang diberikan (1-5)
# 'at': Tanggal ulasan dibuat

- **Metode yang digunakan:** *Slicing* (pemotongan) DataFrame.
- **Alasan penggunaan:** Efisiensi dan Relevansi. Tidak semua kolom yang disediakan oleh scraper (seperti `reviewId`, `userImage`, `replyContent`, dll.) diperlukan untuk analisis sentimen. Dengan memilih hanya kolom-kolom esensial (`userName`, `content`, `score`, `at`), kita:
    1.  Mengurangi penggunaan memori.
    2.  Menyederhanakan dataset, membuatnya lebih mudah untuk ditangani.
    3.  Fokus pada fitur-fitur yang akan digunakan langsung: `content` untuk teks, `score` untuk pelabelan, `userName` dan `at` untuk analisis kontekstual jika diperlukan.
- **Insight dan Hasil yang didapat:** Proses ini mengubah DataFrame mentah yang "gemuk" menjadi DataFrame yang "ramping" dan siap pakai, tanpa kehilangan informasi yang krusial untuk proyek.

In [7]:
# --- Langkah 5: Simpan DataFrame ke File CSV ---
# Menyimpan data bersih ke dalam file CSV.
output_filename = 'dataset_mamikos.csv' # meberikan nama output file 'dataset_mamikos.csv'
df_final.to_csv(output_filename, index=False)

print(f"\nData ulasan (sampel {len(df_final)} baris) telah berhasil disimpan ke file: {output_filename}")


Data ulasan (sampel 10300 baris) telah berhasil disimpan ke file: dataset_mamikos.csv


**Output:**
`Data ulasan (sampel 10.300 baris) telah berhasil disimpan ke file: dataset_mamikos.csv`

- **Metode yang digunakan:** Menyimpan DataFrame ke file menggunakan metode `.to_csv()`.
- **Alasan penggunaan:** **Persistensi Data.** Tahap ini sangat penting untuk memisahkan alur kerja. Dengan menyimpan hasil scraping ke file `.csv`, kita tidak perlu menjalankan ulang proses scraping yang memakan waktu setiap kali kita ingin mengerjakan notebook pemodelan. Notebook pemodelan nantinya hanya perlu memuat file `dataset_mamikos.csv` ini, membuat proses pengembangan lebih cepat dan modular. Parameter `index=False` digunakan untuk mencegah Pandas menulis kolom indeks yang tidak perlu ke dalam file.
- **Insight dan Hasil yang didapat:** Kengonfirmasi bahwa aset data utama proyek telah berhasil dibuat dan disimpan. Proyek ini sekarang memiliki sumber data yang solid dan dapat digunakan kembali.

In [8]:
# Menampilkan 20 baris pertama dari data yang disimpan
print("\nContoh 20 data pertama dari file yang disimpan:")
print(df_final.head(20))


Contoh 20 data pertama dari file yang disimpan:
              userName                                            content  \
1350   Pengguna Google                                          la respon   
1082   Pengguna Google  Bagi siapa saja yang membutuhkan sewa kos kosa...   
673    Pengguna Google                                            bagus..   
3043   Pengguna Google                  Aplikasi yang sangat membantu👍👍👍👍   
10296  Pengguna Google                                           Gandosss   
2275   Pengguna Google  Aplikasi yang sangat bagus sangat membantu bag...   
5769   Pengguna Google                                              bagus   
5958   Pengguna Google                                  sangat bermanfaat   
542    Pengguna Google  Baru Pasang kost sekitar 2mingguan udah njepra...   
6018   Pengguna Google  Awalnya ragu untuk cari kost ditempat yg gak a...   
7787   Pengguna Google  Memudahkan pencarian, sebaiknya dipersering up...   
35     Pengguna Google  san

- **Metode yang digunakan:** Inspeksi data menggunakan metode `.head()`.
- **Alasan penggunaan:** Sebagai langkah verifikasi akhir atau *sanity check*. Ini memungkinkan kita untuk melihat secara langsung sampel dari data yang telah disimpan, memastikan bahwa struktur kolom, tipe data, dan isinya sesuai dengan yang diharapkan.
- **Insight dan Hasil yang didapat:**
  - Menunjukkan data yang bersih dan terstruktur dengan benar dalam 4 kolom yang telah dipilih.
  - Terdapat juga variasi skor yang akan menjadi dasar pelabelan sentimen (Negatif, Netral, Positif).
  - Format tanggal pada kolom `at` terlihat konsisten. Secara keseluruhan, data ini siap untuk tahap selanjutnya yaitu preprocessing dan pemodelan.